TSAN detecting a race related to logging

Error as reported by the sanitizer:

```
E         WARNING: ThreadSanitizer: data race (pid=66)
E           Read of size 8 at 0x7f2f5aeaa048 by thread T4 (mutexes: read M2102):
E             #0 std::__1::ios_base::width() const /opt/llvm/bin/../include/c++/v1/ios:523:12 (nighthawk_service+0x3286229)
E             #1 absl::operator<<(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, absl::string_view) /proc/self/cwd/external/com_google_absl/absl/strings/string_view.cc:65:31 (nighthawk_service+0x870710e)
E             #2 Envoy::Logger::StderrSinkDelegate::log(absl::string_view) /proc/self/cwd/external/envoy/source/common/common/logger.cc:55:13 (nighthawk_service+0x7a4797e)
E             #3 Envoy::Logger::DelegatingLogSink::log(spdlog::details::log_msg const&) /proc/self/cwd/external/envoy/source/common/common/logger.cc:91:12 (nighthawk_service+0x7a47f39)
E             #4 spdlog::logger::sink_it_(spdlog::details::log_msg const&) /proc/self/cwd/external/com_github_gabime_spdlog/include/spdlog/logger-inl.h:186:23 (nighthawk_service+0x327a2eb)
E             #5 spdlog::logger::log_it_(spdlog::details::log_msg const&, bool, bool) /proc/self/cwd/external/com_github_gabime_spdlog/include/spdlog/logger-inl.h:170:9 (nighthawk_service+0x3345e1a)
E             #6 void spdlog::logger::log_<fmt::v7::basic_string_view<char>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/external/com_github_gabime_spdlog/include/spdlog/logger.h:342:13 (nighthawk_service+0x335f7b9)
E             #7 void spdlog::logger::log<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/external/com_github_gabime_spdlog/include/spdlog/logger.h:87:9 (nighthawk_service+0x3323371)
E             #8 Nighthawk::Client::RequestSourceServiceImpl::RequestStream(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*) /proc/self/cwd/source/client/service_impl.cc:142:5 (nighthawk_service+0x331bbfa)
E             #9 nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0::operator()(nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*) const /proc/self/cwd/bazel-out/k8-dbg/bin/api/request_source/service.grpc.pb.cc:64:32 (nighthawk_service+0x795b581)
E             #10 decltype(std::__1::forward<nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0&>(fp)(std::__1::forward<nighthawk::request_source::NighthawkRequestSourceService::Service*>(fp0), std::__1::forward<grpc::ServerContext*>(fp0), std::__1::forward<grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*>(fp0))) std::__1::__invoke<nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0&, nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*>(nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0&, nighthawk::request_source::NighthawkRequestSourceService::Service*&&, grpc::ServerContext*&&, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*&&) /opt/llvm/bin/../include/c++/v1/type_traits:3899:1 (nighthawk_service+0x795b45d)
E             #11 grpc::Status std::__1::__invoke_void_return_wrapper<grpc::Status>::__call<nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0&, nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*>(nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0&, nighthawk::request_source::NighthawkRequestSourceService::Service*&&, grpc::ServerContext*&&, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*&&) /opt/llvm/bin/../include/c++/v1/__functional_base:317:16 (nighthawk_service+0x795b311)
E             #12 std::__1::__function::__alloc_func<nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0, std::__1::allocator<nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0>, grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>::operator()(nighthawk::request_source::NighthawkRequestSourceService::Service*&&, grpc::ServerContext*&&, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*&&) /opt/llvm/bin/../include/c++/v1/functional:1557:16 (nighthawk_service+0x795b241)
E             #13 std::__1::__function::__func<nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0, std::__1::allocator<nighthawk::request_source::NighthawkRequestSourceService::Service::Service()::$_0>, grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>::operator()(nighthawk::request_source::NighthawkRequestSourceService::Service*&&, grpc::ServerContext*&&, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*&&) /opt/llvm/bin/../include/c++/v1/functional:1731:12 (nighthawk_service+0x7959540)
E             #14 std::__1::__function::__value_func<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>::operator()(nighthawk::request_source::NighthawkRequestSourceService::Service*&&, grpc::ServerContext*&&, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*&&) const /opt/llvm/bin/../include/c++/v1/functional:1884:16 (nighthawk_service+0x797835a)
E             #15 std::__1::function<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>::operator()(nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*) const /opt/llvm/bin/../include/c++/v1/functional:2556:12 (nighthawk_service+0x7978212)
E             #16 grpc::internal::BidiStreamingHandler<nighthawk::request_source::NighthawkRequestSourceService::Service, nighthawk::request_source::RequestStreamRequest, nighthawk::request_source::RequestStreamResponse>::BidiStreamingHandler(std::__1::function<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, nighthawk::request_source::NighthawkRequestSourceService::Service*)::'lambda'(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)::operator()(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*) const /proc/self/cwd/external/com_github_grpc_grpc/include/grpcpp/impl/codegen/method_handler.h:321:22 (nighthawk_service+0x797810d)
E             #17 decltype(std::__1::forward<nighthawk::request_source::NighthawkRequestSourceService::Service>(fp)(std::__1::forward<nighthawk::request_source::RequestStreamRequest>(fp0)...)) std::__1::__invoke<grpc::internal::BidiStreamingHandler<nighthawk::request_source::NighthawkRequestSourceService::Service, nighthawk::request_source::RequestStreamRequest, nighthawk::request_source::RequestStreamResponse>::BidiStreamingHandler(std::__1::function<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, nighthawk::request_source::NighthawkRequestSourceService::Service*)::'lambda'(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)&, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*>(nighthawk::request_source::NighthawkRequestSourceService::Service&&, nighthawk::request_source::RequestStreamRequest&&...) /opt/llvm/bin/../include/c++/v1/type_traits:3899:1 (nighthawk_service+0x7978017)
E             #18 grpc::Status std::__1::__invoke_void_return_wrapper<grpc::Status>::__call<grpc::internal::BidiStreamingHandler<nighthawk::request_source::NighthawkRequestSourceService::Service, nighthawk::request_source::RequestStreamRequest, nighthawk::request_source::RequestStreamResponse>::BidiStreamingHandler(std::__1::function<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, nighthawk::request_source::NighthawkRequestSourceService::Service*)::'lambda'(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)&, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*>(nighthawk::request_source::NighthawkRequestSourceService::Service&&...) /opt/llvm/bin/../include/c++/v1/__functional_base:317:16 (nighthawk_service+0x7977f01)
E             #19 std::__1::__function::__alloc_func<grpc::internal::BidiStreamingHandler<nighthawk::request_source::NighthawkRequestSourceService::Service, nighthawk::request_source::RequestStreamRequest, nighthawk::request_source::RequestStreamResponse>::BidiStreamingHandler(std::__1::function<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, nighthawk::request_source::NighthawkRequestSourceService::Service*)::'lambda'(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*), std::__1::allocator<grpc::internal::BidiStreamingHandler<nighthawk::request_source::NighthawkRequestSourceService::Service, nighthawk::request_source::RequestStreamRequest, nighthawk::request_source::RequestStreamResponse>::BidiStreamingHandler(std::__1::function<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, nighthawk::request_source::NighthawkRequestSourceService::Service*)::'lambda'(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, grpc::Status (grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>::operator()(grpc::ServerContext*&&, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*&&) /opt/llvm/bin/../include/c++/v1/functional:1557:16 (nighthawk_service+0x7977e51)
E             #20 std::__1::__function::__func<grpc::internal::BidiStreamingHandler<nighthawk::request_source::NighthawkRequestSourceService::Service, nighthawk::request_source::RequestStreamRequest, nighthawk::request_source::RequestStreamResponse>::BidiStreamingHandler(std::__1::function<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, nighthawk::request_source::NighthawkRequestSourceService::Service*)::'lambda'(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*), std::__1::allocator<grpc::internal::BidiStreamingHandler<nighthawk::request_source::NighthawkRequestSourceService::Service, nighthawk::request_source::RequestStreamRequest, nighthawk::request_source::RequestStreamResponse>::BidiStreamingHandler(std::__1::function<grpc::Status (nighthawk::request_source::NighthawkRequestSourceService::Service*, grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, nighthawk::request_source::NighthawkRequestSourceService::Service*)::'lambda'(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>, grpc::Status (grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>::operator()(grpc::ServerContext*&&, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*&&) /opt/llvm/bin/../include/c++/v1/functional:1731:12 (nighthawk_service+0x7975f13)
E             #21 std::__1::__function::__value_func<grpc::Status (grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>::operator()(grpc::ServerContext*&&, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*&&) const /opt/llvm/bin/../include/c++/v1/functional:1884:16 (nighthawk_service+0x797d0e4)
E             #22 std::__1::function<grpc::Status (grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*)>::operator()(grpc::ServerContext*, grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>*) const /opt/llvm/bin/../include/c++/v1/functional:2556:12 (nighthawk_service+0x797cfcf)
E             #23 grpc::internal::TemplatedBidiStreamingHandler<grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>, false>::RunHandler(grpc::internal::MethodHandler::HandlerParameter const&)::'lambda'()::operator()() const /proc/self/cwd/external/com_github_grpc_grpc/include/grpcpp/impl/codegen/method_handler.h:272:14 (nighthawk_service+0x797cefd)
E             #24 grpc::Status grpc::internal::CatchingFunctionHandler<grpc::internal::TemplatedBidiStreamingHandler<grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>, false>::RunHandler(grpc::internal::MethodHandler::HandlerParameter const&)::'lambda'()>(grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>&&) /proc/self/cwd/external/com_github_grpc_grpc/include/grpcpp/impl/codegen/method_handler.h:42:12 (nighthawk_service+0x7978a01)
E             #25 grpc::internal::TemplatedBidiStreamingHandler<grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse, nighthawk::request_source::RequestStreamRequest>, false>::RunHandler(grpc::internal::MethodHandler::HandlerParameter const&) /proc/self/cwd/external/com_github_grpc_grpc/include/grpcpp/impl/codegen/method_handler.h:271:29 (nighthawk_service+0x797480f)
E             #26 grpc::Server::SyncRequest::CallData::ContinueRunAfterInterception() /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/server/server_cc.cc:489:18 (nighthawk_service+0x3e694ec)
E             #27 grpc::Server::SyncRequest::CallData::Run(std::__1::shared_ptr<grpc::Server::GlobalCallbacks> const&, bool) /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/server/server_cc.cc:476:9 (nighthawk_service+0x3e68e44)
E             #28 grpc::Server::SyncRequestThreadManager::DoWork(void*, bool, bool) /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/server/server_cc.cc:806:11 (nighthawk_service+0x3e66abe)
E             #29 grpc::ThreadManager::MainWorkLoop() /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/thread_manager/thread_manager.cc:213:9 (nighthawk_service+0x3ece0b2)
E             #30 grpc::ThreadManager::WorkerThread::Run() /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/thread_manager/thread_manager.cc:43:13 (nighthawk_service+0x3ecdb7b)
E             #31 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::operator()(void*) const /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/thread_manager/thread_manager.cc:35:69 (nighthawk_service+0x3ecf054)
E             #32 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::__invoke(void*) /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/thread_manager/thread_manager.cc:35:7 (nighthawk_service+0x3eceff8)
E             #33 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const /proc/self/cwd/external/com_github_grpc_grpc/src/core/lib/gprpp/thd_posix.cc:140:23 (nighthawk_service+0x4484a5c)
E             #34 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) /proc/self/cwd/external/com_github_grpc_grpc/src/core/lib/gprpp/thd_posix.cc:110:21 (nighthawk_service+0x4484848)
E
E           Previous write of size 8 at 0x7f2f5aeaa048 by main thread (mutexes: read M2102):
E             #0 std::__1::ios_base::width(long) /opt/llvm/bin/../include/c++/v1/ios:531:14 (nighthawk_service+0x3286447)
E             #1 absl::operator<<(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, absl::string_view) /proc/self/cwd/external/com_google_absl/absl/strings/string_view.cc:76:7 (nighthawk_service+0x8707326)
E             #2 Envoy::Logger::StderrSinkDelegate::log(absl::string_view) /proc/self/cwd/external/envoy/source/common/common/logger.cc:55:13 (nighthawk_service+0x7a4797e)
E             #3 Envoy::Logger::DelegatingLogSink::log(spdlog::details::log_msg const&) /proc/self/cwd/external/envoy/source/common/common/logger.cc:91:12 (nighthawk_service+0x7a47f39)
E             #4 spdlog::logger::sink_it_(spdlog::details::log_msg const&) /proc/self/cwd/external/com_github_gabime_spdlog/include/spdlog/logger-inl.h:186:23 (nighthawk_service+0x327a2eb)
E             #5 spdlog::logger::log_it_(spdlog::details::log_msg const&, bool, bool) /proc/self/cwd/external/com_github_gabime_spdlog/include/spdlog/logger-inl.h:170:9 (nighthawk_service+0x3345e1a)
E             #6 void spdlog::logger::log_<fmt::v7::basic_string_view<char>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/external/com_github_gabime_spdlog/include/spdlog/logger.h:342:13 (nighthawk_service+0x335f7b9)
E             #7 void spdlog::logger::log<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/external/com_github_gabime_spdlog/include/spdlog/logger.h:87:9 (nighthawk_service+0x3323371)
E             #8 Nighthawk::Client::ServiceMain::start() /proc/self/cwd/source/client/service_main.cc:75:3 (nighthawk_service+0x338398f)
E             #9 main /proc/self/cwd/source/exe/service_main_entry.cc:19:13 (nighthawk_service+0x3276069)
E           Location is global 'std::__1::cerr' of size 160 at 0x7f2f5aeaa028 (libc++.so.1+0x0000002c5048)
E
E           Mutex M2102 (0x7b1000016c18) created at:
E             #0 __tsan_mutex_create /local/mnt/workspace/bcain_clang_bcain-ubuntu_7561/final/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interface_ann.cpp:466:3 (nighthawk_service+0x3250b59)
E             #1 absl::Mutex::Mutex() /proc/self/cwd/external/com_google_absl/absl/synchronization/mutex.h:937:3 (nighthawk_service+0x3354011)
E             #2 Envoy::Logger::DelegatingLogSink::DelegatingLogSink() /proc/self/cwd/external/envoy/source/common/common/logger.h:201:3 (nighthawk_service+0x7a4da7d)
E             #3 Envoy::Logger::DelegatingLogSink::init() /proc/self/cwd/external/envoy/source/common/common/logger.cc:107:50 (nighthawk_service+0x7a483e1)
E             #4 Envoy::Logger::Registry::getSink() /proc/self/cwd/external/envoy/source/common/common/logger.h:284:46 (nighthawk_service+0x38ed63d)
E             #5 Envoy::Logger::StandardLogger::StandardLogger(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/external/envoy/source/common/common/logger.cc:22:53 (nighthawk_service+0x7a472ea)
E             #6 Envoy::Logger::Registry::allLoggers() /proc/self/cwd/external/envoy/source/common/common/logger.cc:192:32 (nighthawk_service+0x7a4954c)
E             #7 Envoy::Logger::Registry::getLog(Envoy::Logger::Id) /proc/self/cwd/external/envoy/source/common/common/logger.cc:196:51 (nighthawk_service+0x7a4ccd1)
E             #8 Envoy::Logger::Loggable<(Envoy::Logger::Id)28>::__log_do_not_use_read_comment() /proc/self/cwd/external/envoy/source/common/common/logger.h:328:39 (nighthawk_service+0x3323405)
E             #9 Nighthawk::Client::ServiceMain::ServiceMain(int, char const**) /proc/self/cwd/source/client/service_main.cc:51:3 (nighthawk_service+0x3382fcf)
E             #10 main /proc/self/cwd/source/exe/service_main_entry.cc:18:17 (nighthawk_service+0x3276058)
E
E           Thread T4 'grpcpp_sync_ser' (tid=77, running) created by main thread at:
E             #0 pthread_create /local/mnt/workspace/bcain_clang_bcain-ubuntu_7561/final/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:966:3 (nighthawk_service+0x320782b)
E             #1 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/external/com_github_grpc_grpc/src/core/lib/gprpp/thd_posix.cc:108:17 (nighthawk_service+0x448438e)
E             #2 grpc_core::Thread::Thread(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/external/com_github_grpc_grpc/src/core/lib/gprpp/thd_posix.cc:186:15 (nighthawk_service+0x4483c66)
E             #3 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*) /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/thread_manager/thread_manager.cc:33:10 (nighthawk_service+0x3ecda07)
E             #4 grpc::ThreadManager::Initialize() /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/thread_manager/thread_manager.cc:143:32 (nighthawk_service+0x3eceee2)
E             #5 grpc::Server::SyncRequestThreadManager::Start() /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/server/server_cc.cc:859:7 (nighthawk_service+0x3e63682)
E             #6 grpc::Server::Start(grpc::ServerCompletionQueue**, unsigned long) /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/server/server_cc.cc:1200:12 (nighthawk_service+0x3e55c6b)
E             #7 grpc::ServerBuilder::BuildAndStart() /proc/self/cwd/external/com_github_grpc_grpc/src/cpp/server/server_builder.cc:411:11 (nighthawk_service+0x3e31880)
E             #8 Nighthawk::Client::ServiceMain::start() /proc/self/cwd/source/client/service_main.cc:67:22 (nighthawk_service+0x3383495)
E             #9 main /proc/self/cwd/source/exe/service_main_entry.cc:19:13 (nighthawk_service+0x3276069)
E
E         SUMMARY: ThreadSanitizer: data race /opt/llvm/bin/../include/c++/v1/ios:523:12 in std::__1::ios_base::width() const
E         ==================
E         [2021-06-08 05:14:21.816][77][trace][main] [source/client/service_impl.cc:142] Inbound RequestStreamRequest quantity: 10
E
E         [2021-06-08 05:14:22.667][77][debug][runtime] [external/envoy/source/common/runtime/runtime_features.cc:31] Unable to use runtime singleton for feature envoy.http.headermap.lazy_map_min_size
E         [2021-06-08 05:14:22.669][77][debug][runtime] [external/envoy/source/common/runtime/runtime_features.cc:20] Unable to use runtime singleton for feature envoy.reloadable_features.header_map_correctly_coalesce_cookies
E         [2021-06-08 05:14:22.731][77][trace][main] [source/client/service_impl.cc:142] Inbound RequestStreamRequest quantity: 10
E
E         [2021-06-08 05:14:22.731][77][debug][runtime] [external/envoy/source/common/runtime/runtime_features.cc:31] Unable to use runtime singleton for feature envoy.http.headermap.lazy_map_min_size
E         [2021-06-08 05:14:22.731][77][debug][runtime] [external/envoy/source/common/runtime/runtime_features.cc:20] Unable to use runtime singleton for feature envoy.reloadable_features.header_map_correctly_coalesce_cookies
E         [2021-06-08 05:14:32.350][77][trace][main] [source/client/service_impl.cc:173] Finishing stream
E         [2021-06-08 05:14:32.415][66][info][main] [source/client/service_main.cc:90] Nighthawk grpc service exits
E         ThreadSanitizer: reported 1 warnings
```

--- a/source/client/service_impl.h
+++ b/source/client/service_impl.h
@@ -81,12 +81,22 @@ class RequestSourceServiceImpl final
       public Envoy::Logger::Loggable<Envoy::Logger::Id::main> {
 
 public:
+  /**
+   * Constructs a new RequestSourceServiceImpl instance.
+   */
+  RequestSourceServiceImpl() {
+    logging_context_ = std::make_unique<Envoy::Logger::Context>(
+        spdlog::level::from_str("info"), "[%T.%f][%t][%L] %v", log_lock_, false);
+  }
+
   grpc::Status RequestStream(
       grpc::ServerContext* context,
       grpc::ServerReaderWriter<nighthawk::request_source::RequestStreamResponse,
                                nighthawk::request_source::RequestStreamRequest>* stream) override;
 
 private:
+  std::unique_ptr<Envoy::Logger::Context> logging_context_;
+  Envoy::Thread::MutexBasicLockable log_lock_;
   RequestSourcePtr createStaticEmptyRequestSource(const uint32_t amount);
 };
 

--- a/test/integration/integration_test_fixtures.py
+++ b/test/integration/integration_test_fixtures.py
@@ -138,7 +138,8 @@ class IntegrationTestBase():
       caplog: The pytest `caplog` test fixture used to examine logged messages.
     """
     if self.grpc_service is not None:
-      assert (self.grpc_service.stop() == 0)
+      if self.grpc_service.stop() != 0:
+        pytest.fail("the Nighthawk GRPC service reported a non-zero exit code when stopped, log lines:\n{}".format('\n'.join(self.grpc_service.log_lines)))
 
     any_failed = False
     for test_server in self._test_servers:

--- a/test/integration/integration_test_fixtures.py
+++ b/test/integration/integration_test_fixtures.py
@@ -139,7 +139,9 @@ class IntegrationTestBase():
     """
     if self.grpc_service is not None:
       if self.grpc_service.stop() != 0:
-        pytest.fail("the Nighthawk GRPC service reported a non-zero exit code when stopped, log lines:\n{}".format('\n'.join(self.grpc_service.log_lines)))
+        pytest.fail(
+            "the Nighthawk GRPC service reported a non-zero exit code when stopped, log lines:\n{}".
+            format('\n'.join(self.grpc_service.log_lines)))
 
     any_failed = False
     for test_server in self._test_servers:

